### ISSUE TITLE
Support automatically inserting missing semicolons during save/formatting

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#45427]
This came up at https://github.com/Dart-Code/Dart-Code/issues/3220 (and also [in dart_style](https://github.com/dart-lang/dart_style/issues/948)). The server seems the correct place for it (given the editor doesn't have enough info, and it's not strictly formatting), so I'm moving it here as a central place for people to find/express interest/add 👍 s.

It would likely rely on the parser recovering well for code without semicolons, and create edits to insert them where they were reported missing. @bwilkerson's suggestion for testing how well that works:

> We could test it by taking some representative samples of code, and for each file remove varying numbers of semicolons, parse the file, and record the number of times a missing semicolon diagnostic was reported in the correct location and the number of times it was either not reported or reported in the wrong location.

**Comment by vsmenon** [comment]
Also, potential a dartfmt request - @munificent

**Comment by lrhn** [comment]
It's non-trivial to determine where semicolons are missing, as opposed to some other error.
For example:
```dart
  int y = 0;
  int x y = 42;
```
reports that a `;` is expected at the `y` after `x`. The real mistake was probably not removing the `y`.

**Comment by diosney** [comment]
@lrhn You're right, but at least it should recover from trivial cases, like line endings like:

```dart
  int y = 0;
  int x = 42
```

It is so QoL, that I'm very accustomed in other languages to reduce typing to let the IDE autocompletes them for me (lazy, I know, sorry for that), and sometimes even I write several lines and in one Reformat sweep the IDE do that and feels so nice that now with Dart I miss it a lot.

**Comment by bernaferrari** [comment]
Prettier does this with EXCELLENCE. It might be useful to see what it solves/what it doesn't. But it is extremely helpful. Very very very helpful. I write code without ; and let it add. In Dart it is always a paint to remember the ;.

**Comment by munificent** [comment]
> Prettier does this with EXCELLENCE. It might be useful to see what it solves/what it doesn't.

This is because semicolons are optional in JavaScript, so a source file missing semicolons is still a completely well-specified JavaScript program that Prettier can thus parse using a standard JS parser.

But JavaScript itself takes on a *ton* of complexity in the language in order to make semicolons optional in the way it does. That's absolutely not an approach we would want to take for Dart.

**Comment by NatoBoram** [comment]
Another way to go about it would be to [make semicolons optional](https://github.com/dart-lang/language/issues/69) and have them auto-inserted at compile time

**Comment by bernaferrari** [comment]
lol, but that's a 10x harder issue, and even in TypeScript, where it is optional, I like that the tooling adds the ; for me. So it wouldn't solve that.
